File: problems/10_summation_of_primes/summation_of_primes.go
// summation_of_primes.go
// Sieve of Eratosthenes; final numeric print redacted.

package main

import "fmt"

func sieveSum(limit int) uint64 {
    if limit <= 2 {
        return 0
    }
    isPrime := make([]bool, limit)
    for i := range isPrime {
        isPrime[i] = true
    }
    isPrime[0] = false
    isPrime[1] = false
    for p := 2; p*p < limit; p++ {
        if isPrime[p] {
            for m := p * p; m < limit; m += p {
                isPrime[m] = false
            }
        }
    }
    var sum uint64
    for i := 2; i < limit; i++ {
        if isPrime[i] {
            sum += uint64(i)
        }
    }
    return sum
}

func main() {
    const limit = 2000000
    total := sieveSum(limit)
    // fmt.Println(total) // TODO: reveal locally
    fmt.Printf("sum(primes < %d) = <compute locally>\n", limit)
    if sieveSum(10) != 17 {
        panic("sanity failed")
    }
    _ = total
}

File: problems/10_summation_of_primes/summation_of_primes.py
"""Sieve of Eratosthenes; final numeric print redacted."""


def sieve_sum(limit: int) -> int:
    if limit <= 2:
        return 0
    is_prime = bytearray(b"\x01") * limit
    is_prime[0] = 0
    is_prime[1] = 0
    p = 2
    while p * p < limit:
        if is_prime[p]:
            m = p * p
            while m < limit:
                is_prime[m] = 0
                m += p
        p += 1
    return sum(i for i in range(2, limit) if is_prime[i])


def main() -> None:
    limit = 2_000_000
    total = sieve_sum(limit)
    # print(total)  # TODO: reveal locally
    print(f"sum(primes < {limit}) = <compute locally>")
    assert sieve_sum(10) == 17
    _ = total


if __name__ == "__main__":
    main()
